utilities/blob_db/blob_db_impl.cc

```diff
@@ -0,0 +1,2073 @@
+  }
+
+  bfile->file_size_ += blob_log::BlobLogRecord::kFooterSize;
+  last_period_write_ += blob_log::BlobLogRecord::kFooterSize;
+  total_blob_space_ += blob_log::BlobLogRecord::kFooterSize;
+  return s;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+//
+////////////////////////////////////////////////////////////////////////////////
+std::vector<Status> BlobDBImpl::MultiGet( const ReadOptions& options,
+  const std::vector<ColumnFamilyHandle*>& column_family,
+  const std::vector<Slice>& keys, std::vector<std::string>* values) {
+
+  auto statuses = db_->MultiGet(options, column_family, keys, values);
+
+  for (size_t i = 0; i < keys.size(); ++i) {
+    if (!statuses[i].ok())
+      continue;
+
+    BlockHandle handle;
+    Slice index_entry_slice((*values)[i]);
+    uint64_t file_number;
+    if (!GetVarint64(&index_entry_slice, &file_number)) {
+      statuses[i] = Status::Corruption();
+      continue;
+    }
+
+    statuses[i] = handle.DecodeFrom(&index_entry_slice);
+    if (!statuses[i].ok())
+      continue;
+
+    std::shared_ptr<BlobFile> bfile;
+    {
+      ReadLock l(&mutex_);
+      auto hitr = blob_files_.find(file_number);
+
+      // file was deleted
+      if (hitr == blob_files_.end()) {
+        statuses[i] = Status::NotFound("Blob Not Found");
+        continue;
+      }
+
+      bfile = hitr->second;
+    }
+
+    std::shared_ptr<RandomAccessFileReader> reader;
+    {
+      WriteLock lockbfile_w(&bfile->mutex_);
+      reader = openRandomAccess_locked(bfile, myenv_, env_options_);
+    }
+
+    // allocate the buffer
+    (*values)[i].resize(handle.size());
+    char *buffer = &((*values)[i])[0];
+
+    Slice blob_value;
+    statuses[i] = reader->Read(handle.offset(), handle.size(), &blob_value, buffer);
+  }
+  return statuses;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+//
+////////////////////////////////////////////////////////////////////////////////
+Status BlobDBImpl::Get(const ReadOptions& options,
+  ColumnFamilyHandle* column_family, const Slice& key, std::string* value) {
+
+  Status s;
+  std::string index_entry;
+  s = db_->Get(options, column_family, key, &index_entry);
+  if (!s.ok())
+    return s;
+
+  BlockHandle handle;
+  Slice index_entry_slice(index_entry);
+  uint64_t file_number;
+  if (!GetVarint64(&index_entry_slice, &file_number))
+    return Status::Corruption();
+
+  s = handle.DecodeFrom(&index_entry_slice);
+  if (!s.ok())
+    return s;
+
+  std::shared_ptr<BlobFile> bfile;
+  {
+    ReadLock l(&mutex_);
+    auto hitr = blob_files_.find(file_number);
+
+    // file was deleted
+    if (hitr == blob_files_.end()) {
+      return Status::NotFound("Blob Not Found");
+    }
+
+    bfile = hitr->second;
+  }
+
+  std::shared_ptr<RandomAccessFileReader> reader;
+  {
+    WriteLock lockbfile_w(&bfile->mutex_);
+    reader = openRandomAccess_locked(bfile, myenv_, env_options_);
```

Comment:
Get will only open a file if it is the first time, or if ulimit has forced the closing of
file descriptors. openRandomAccess_locked will typically not have to open a file,
if it has been previously opened.